apollo-federation/src/operation/optimize.rs

```diff
@@ -894,19 +891,25 @@ impl SelectionSet {
                     ignore_missing_typename: true,
                 },
             );
-            if matches!(res, Containment::NotContained) {
-                continue; // Not eligible; Skip it.
-            }
-            if matches!(res, Containment::Equal) && full_match_condition.check(&candidate) {
-                // Special case: Found a fragment that covers the full selection set.
-                return Ok(candidate.into());
-            }
-            // Note that if a fragment applies to only a subset of the sub-selections, then we
-            // really only can use it if that fragment is defined _without_ directives.
-            if !candidate.directives.is_empty() {
-                continue; // Not eligible as a partial selection; Skip it.
+            match res {
+                Containment::Equal if full_match_condition.check(&candidate) => {
+                    if !validator.check_can_reuse_fragment_and_track_it(&at_type)? {
```

Comment:
This is the big difference: `check_can_reuse_fragment` is where `do_merge_with` is called, potentially a lot, and it is potentially expensive for big selection sets. JS only calls it in this rare condition, while we were calling it all the time (and in another spot that I still need to add in Rust).